src/service/url-replacements-impl.js

```diff
@@ -197,6 +210,24 @@ export class UrlReplacements {
         defaultValue;
     });
 
+    this.setAsync_('QUERY_PARAM', (param, defaultValue = '') => {
+      user().assert(param,
+          'The first argument to QUERY_PARAM, the query string ' +
+          'param is required');
+      let promise = trackImpressionPromise;
+      if (!promise) {
+        promise = Promise.resolve();
+      }
+      return promise.then(() => {
+        const url = parseUrl(this.ampdoc.win.location.href);
```

Comment:
I added a `getQueryParamData_` function. I think you have your point, it definitely reduce repeated code. but adding an extra method else seems unclear to me actually.